keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,85 @@
+import io.improbable.keanu.algorithms.ProbabilisticModel;
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.graphtraversal.TopologicalSort;
+import io.improbable.keanu.algorithms.mcmc.NetworkSamplesGenerator;
+import io.improbable.keanu.algorithms.mcmc.SamplingAlgorithm;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.util.status.StatusBar;
+import io.improbable.keanu.vertices.Vertex;
+import org.nd4j.base.Preconditions;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class Forward implements PosteriorSamplingAlgorithm {
+
+    private final KeanuRandom random;
+
+    public Forward() {
+        this(KeanuRandom.getDefaultRandom());
+    }
+
+    public Forward(KeanuRandom random) {
+        this.random = random;
+    }
+
+    /**
+     * Samples from the prior of a Probabilistic Model.
+     *
+     * Samples are taken by sampling from the prior of the desired variables in topological order.
+     *
+     * @param model the model to sample from
+     * @param variablesToSampleFrom the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        for (Variable variable : variablesToSampleFrom) {
+            Preconditions.checkArgument(variable instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex. Received : " + variable);
+        }
+
+        Variable latent = latentVariables.get(0);
+        BayesianNetwork network = new BayesianNetwork(((Vertex) latent).getConnectedGraph());
+
+        List<Vertex> observedVertices = network.getObservedVertices();
+        checkUpstreamOfObservedDoesNotContainProbabilistic(observedVertices);
+
+        List<Vertex> verticesToSampleFrom = variablesToSampleFrom.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+        Set<Vertex> allDownstreamVertices = allDownstreamVertices(network.getLatentVertices());
```

Comment:
As discussed offline:
1) I don't think Lambda Sections always work here as we need to re-evaluate every node in the tree back to the root. (We'd be looking for a transitive closure kind of thing)
2) For the set of vertices we actually need to re-evaluate we want the intersection of (downstream of all latents) & (upstream from the nodes we're sampling from).